node/pkg/watchers/solana/shim.go

```diff
@@ -0,0 +1,359 @@
+	postMessage, err := shimParsePostMessage(s.shimPostMessageDiscriminator, innerInstructions[startIdx].Data)
+	if err != nil {
+		return false, fmt.Errorf("failed to parse inner shim post message instruction %d, %d: %w", outerIdx, startIdx, err)
+	}
+
+	if postMessage == nil {
+		return false, nil
+	}
+
+	alreadyProcessed.add(outerIdx, startIdx)
+
+	err = s.shimProcessRest(logger, whProgramIndex, shimProgramIndex, tx, innerInstructions, outerIdx, startIdx+1, postMessage, alreadyProcessed, isReobservation)
+	if err != nil {
+		return false, fmt.Errorf("failed to process inner instructions for top-level shim instruction %d: %w", outerIdx, err)
+	}
+
+	return true, nil
+}
+
+func (s *SolanaWatcher) shimProcessRest(
+	logger *zap.Logger,
+	whProgramIndex uint16,
+	shimProgramIndex uint16,
+	tx *solana.Transaction,
+	innerInstructions []solana.CompiledInstruction,
+	outerIdx int,
+	startIdx int,
+	postMessage *ShimPostMessageData,
+	alreadyProcessed ShimAlreadyProcessed,
+	isReobservation bool,
+) error {
+	// Loop through the inner instructions after the shim PostMessage and do two things:
+	// 1) Find the core event and verify it is unreliable with an empty payload.
+	// 2) Find the shim MessageEvent to get the rest of the fields we need for the observation.
+
+	var verifiedCoreEvent bool
+	var messageEvent *ShimMessageEventData
+	var err error
+	coreEventFound := false
+	for idx := startIdx; idx < len(innerInstructions); idx++ {
+		inst := innerInstructions[idx]
+		if inst.ProgramIDIndex == whProgramIndex {
+			if verifiedCoreEvent, err = shimVerifyCoreMessage(inst.Data); err != nil {
+				return fmt.Errorf("failed to verify inner core instruction for top-level shim instruction %d: %w", outerIdx, err)
+			}
+			alreadyProcessed.add(outerIdx, idx)
+			coreEventFound = true
+		} else if inst.ProgramIDIndex == shimProgramIndex {
+			if !coreEventFound {
+				return fmt.Errorf("detected an inner shim message event instruction before the core event for top-level shim instruction %d: %w", outerIdx, err)
+			}
+			messageEvent, err = shimParseMessageEvent(s.shimMessageEventDiscriminator, inst.Data)
+			if err != nil {
+				return fmt.Errorf("failed to parse inner shim message event instruction for top-level shim instruction %d: %w", outerIdx, err)
+			}
+			alreadyProcessed.add(outerIdx, idx)
+		}
+
+		if verifiedCoreEvent && messageEvent != nil {
+			break
+		}
+	}
+
+	if !verifiedCoreEvent {
+		return fmt.Errorf("failed to find inner core instruction for top-level shim instruction %d", outerIdx)
+	}
+
+	if messageEvent == nil {
+		return fmt.Errorf("failed to find inner shim message event instruction for top-level shim instruction %d", outerIdx)
+	}
+
+	commitment, err := postMessage.ConsistencyLevel.Commitment()
+	if err != nil {
+		return fmt.Errorf("failed to determine commitment: %w", err)
+	}
+
+	if commitment != s.commitment && (!isReobservation || s.commitment != rpc.CommitmentFinalized) {
```

Comment:
I know this matches https://github.com/wormhole-foundation/wormhole/blob/2615d55a5e59cddb5a9bc9a920c13dc1556323fa/node/pkg/watchers/solana/client.go#L927, but it is confusing to read :sweat:

Reply:
This is being addressed with PR#4231.